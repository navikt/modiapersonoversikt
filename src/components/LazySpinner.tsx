import * as React from 'react';
import styled, { css } from 'styled-components/macro';
import { theme } from '../styles/personOversiktTheme';
import NavFrontendSpinner from 'nav-frontend-spinner';
import FillCenterAndFadeIn from './FillCenterAndFadeIn';
import DelayRender from './DelayRender';

interface Props {
    delay?: number;
    type?: 'XXS' | 'XS' | 'S' | 'M' | 'L' | 'XL' | 'XXL' | 'XXXL';
    padding?: string;
    className?: string;
}

const Styling = styled.span<{ padding?: string }>`
    display: flex;
    justify-content: center;
    ${theme.animation.fadeIn};
    ${props =>
        props.padding
            ? css`
                  padding: ${props.padding};
              `
            : ''};
`;

export function CenteredLazySpinner(props: Props) {
    return (
        <FillCenterAndFadeIn>
            <LazySpinner {...props} />
        </FillCenterAndFadeIn>
    );
}

function LazySpinner(props: Props) {
    return (
        <DelayRender delay={props.delay || 300}>
            <Styling padding={props.padding} className={props.className}>
                <NavFrontendSpinner type={props.type || 'L'} />
            </Styling>
        </DelayRender>
    );
}

export default LazySpinner;
