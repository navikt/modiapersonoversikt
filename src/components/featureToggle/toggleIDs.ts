export enum FeatureToggles {
    VisTilbakemelding = 'vis-tilbakemelding',
    useNewDialogComponents = 'modiapersonoversikt.vis-ny-meldingsvisning'
}
