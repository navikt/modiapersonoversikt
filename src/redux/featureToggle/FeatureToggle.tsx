import * as React from 'react';
import { ReactNode } from 'react';
import { AppState } from '../reducers';
import { AsyncDispatch } from '../ThunkTypes';
import { connect } from 'react-redux';
import { getFetchFeatureToggleAndDispatchToRedux } from './getFeatureToggleAndDispatchToRedux';
import LazySpinner from '../../components/LazySpinner';

export enum DisplayWhenToggleIs {
    ON,
    OFF
}

interface OwnProps {
    children: ReactNode;
    toggleID: string;
    mode: DisplayWhenToggleIs;
}

interface StateProps {
    isOn: boolean;
}

interface DispatchProps {
    fetchFeatureToggle: () => void;
}

type Props = OwnProps & StateProps & DispatchProps;

class FeatureToggle extends React.PureComponent<Props> {
    componentDidMount() {
        this.props.fetchFeatureToggle();
    }

    shouldDisplay() {
        if (this.props.mode === DisplayWhenToggleIs.ON && this.props.isOn) {
            return true;
        }
        if (this.props.mode === DisplayWhenToggleIs.OFF && !this.props.isOn) {
            return true;
        }
        return false;
    }

    render() {
        const pending = this.props.isOn === undefined;
        if (pending) {
            return <LazySpinner type="S" delay={1000}/>;
        }

        if (this.shouldDisplay()) {
            return this.props.children;
        }

        return null;
    }
}

function mapStateToProps(state: AppState, ownProps: OwnProps): StateProps {
    return {
        isOn: state.featureToggle[ownProps.toggleID]
    };
}

function mapDispatchToProps(dispatch: AsyncDispatch, ownProps: OwnProps): DispatchProps {
    return {
        fetchFeatureToggle: () => getFetchFeatureToggleAndDispatchToRedux(ownProps.toggleID, dispatch)
    };
}

export default connect(mapStateToProps, mapDispatchToProps)(FeatureToggle);
