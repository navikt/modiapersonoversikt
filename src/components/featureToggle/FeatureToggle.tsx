import * as React from 'react';
import { ReactNode } from 'react';
import LazySpinner from '../LazySpinner';
import { FeatureToggles } from './toggleIDs';
import featuretoggles from '../../rest/resources/featuretogglesResource';

export enum DisplayWhenToggleIs {
    ON,
    OFF
}

interface Props {
    children: ReactNode;
    toggleID: FeatureToggles;
    mode: DisplayWhenToggleIs;
}

function shouldDisplay(mode: DisplayWhenToggleIs, featureToggleIsOn: boolean): boolean {
    switch (mode) {
        case DisplayWhenToggleIs.ON:
            return featureToggleIsOn;
        case DisplayWhenToggleIs.OFF:
            return !featureToggleIsOn;
    }
}

function FeatureToggle(props: Props) {
    return featuretoggles.useRenderer({
        ifPending: <LazySpinner type="S" delay={1000} />,
        ifData: (toggles) => {
            if (shouldDisplay(props.mode, toggles[props.toggleID])) {
                return <>{props.children}</>;
            } else {
                return null;
            }
        }
    });
}

export default FeatureToggle;
