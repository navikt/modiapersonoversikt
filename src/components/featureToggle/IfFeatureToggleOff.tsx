import * as React from 'react';
import { ReactNode } from 'react';
import FeatureToggle, { DisplayWhenToggleIs } from './FeatureToggle';
import { FeatureToggles } from './toggleIDs';

interface Props {
    children: ReactNode;
    toggleID: FeatureToggles;
}

function IfFeatureToggleOff(props: Props) {
    return <FeatureToggle {...props} mode={DisplayWhenToggleIs.OFF} />;
}

export default IfFeatureToggleOff;
