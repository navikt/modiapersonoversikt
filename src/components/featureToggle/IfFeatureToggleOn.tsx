import * as React from 'react';
import { ReactNode } from 'react';
import FeatureToggle, { DisplayWhenToggleIs } from './FeatureToggle';
import { FeatureToggles } from './toggleIDs';

interface Props {
    children: ReactNode;
    toggleID: FeatureToggles;
}

function IfFeatureToggleOn(props: Props) {
    return <FeatureToggle {...props} mode={DisplayWhenToggleIs.ON} />;
}

export default IfFeatureToggleOn;
