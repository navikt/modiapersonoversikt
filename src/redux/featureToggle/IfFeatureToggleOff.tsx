import * as React from 'react';
import { ReactNode } from 'react';
import FeatureToggle, { DisplayWhenToggleIs } from './FeatureToggle';

interface Props {
    children: ReactNode;
    toggleID: string;
}

function IfFeatureToggleOff(props: Props) {
    return <FeatureToggle {...props} mode={DisplayWhenToggleIs.OFF}/>;
}

export default IfFeatureToggleOff;
