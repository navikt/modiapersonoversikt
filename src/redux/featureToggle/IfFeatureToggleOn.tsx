import * as React from 'react';
import { ReactNode } from 'react';
import FeatureToggle, { DisplayWhenToggleIs } from './FeatureToggle';

interface Props {
    children: ReactNode;
    toggleID: string;
}

function IfFeatureToggleOn(props: Props) {
    return <FeatureToggle {...props} mode={DisplayWhenToggleIs.ON}/>;
}

export default IfFeatureToggleOn;
