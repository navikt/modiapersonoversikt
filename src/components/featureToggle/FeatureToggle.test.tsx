import * as React from 'react';
import { mount } from 'enzyme';
import IfFeatureToggleOff from './IfFeatureToggleOff';
import LazySpinner from '../LazySpinner';
import IfFeatureToggleOn from './IfFeatureToggleOn';
import TestProvider from '../../test/Testprovider';
import FeatureToggle from './FeatureToggle';
import { getTestStore } from '../../setupTests';
import { featureToggleActionNames } from '../../redux/restReducers/featureToggles';
import { FeatureToggles } from './toggleIDs';

const toggleId = FeatureToggles.Tooltip;
const testStore = getTestStore();

beforeEach(() => {
    FeatureToggle.prototype.render = FeatureToggle.prototype.actualRender;
    FeatureToggle.prototype.componentDidMount = () => null;
});

function setToggleTo(value: boolean | undefined) {
    testStore.dispatch({ type: featureToggleActionNames.FINISHED, data: { [toggleId]: value } });
}

test('viser innhold i FeatureToggle dersom feature-toggle er på', () => {
    setToggleTo(true);

    const result = mount(
        <TestProvider customStore={testStore}>
            <IfFeatureToggleOn toggleID={toggleId}>Jeg er på</IfFeatureToggleOn>
            <IfFeatureToggleOff toggleID={toggleId}>Jeg er av</IfFeatureToggleOff>
        </TestProvider>
    );

    expect(result.text()).toContain('Jeg er på');
});

test('viser innhold i IfFeatureToggleOff dersom feature-toggle er av', () => {
    setToggleTo(false);

    const result = mount(
        <TestProvider customStore={testStore}>
            <IfFeatureToggleOn toggleID={toggleId}>Jeg er på</IfFeatureToggleOn>
            <IfFeatureToggleOff toggleID={toggleId}>Jeg er av</IfFeatureToggleOff>
        </TestProvider>
    );

    expect(result.text()).toContain('Jeg er av');
});

test('viser LazySpinner dersom feature-toggle ikke er satt', () => {
    setToggleTo(undefined);

    const result = mount(
        <TestProvider customStore={testStore}>
            <IfFeatureToggleOn toggleID={toggleId}>Jeg er på</IfFeatureToggleOn>
            <IfFeatureToggleOff toggleID={toggleId}>Jeg er av</IfFeatureToggleOff>
        </TestProvider>
    );

    expect(result.contains(<LazySpinner type="S" delay={1000} />)).toBe(true);
});
