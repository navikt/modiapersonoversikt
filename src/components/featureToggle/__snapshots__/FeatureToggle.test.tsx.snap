// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`viser LazySpinner dersom feature-toggle ikke er satt 1`] = `
.c0 {
  overflow: hidden;
  display: flex;
  justify-content: center;
  animation: edHiKw 0.3s ease-out;
}

<div>
  <span
    class="c0"
  >
    <svg
      aria-label="Laster innhold"
      class="spinner spinner--s"
      focusable="false"
      height="32"
      kind="spinner-negativ"
      preserveAspectRatio="xMidYMid"
      viewBox="0 0 50 50"
      width="32"
    >
      <title>
        Venter...
      </title>
      <circle
        cx="25"
        cy="25"
        fill="none"
        r="20"
        stroke="#eee"
        stroke-width="5"
        xmlns="http://www.w3.org/2000/svg"
      />
      <circle
        cx="25"
        cy="25"
        fill="none"
        r="20"
        stroke="#888"
        stroke-dasharray="50 155"
        stroke-linecap="round"
        stroke-width="5"
      />
    </svg>
  </span>
  <span
    class="c0"
  >
    <svg
      aria-label="Laster innhold"
      class="spinner spinner--s"
      focusable="false"
      height="32"
      kind="spinner-negativ"
      preserveAspectRatio="xMidYMid"
      viewBox="0 0 50 50"
      width="32"
    >
      <title>
        Venter...
      </title>
      <circle
        cx="25"
        cy="25"
        fill="none"
        r="20"
        stroke="#eee"
        stroke-width="5"
        xmlns="http://www.w3.org/2000/svg"
      />
      <circle
        cx="25"
        cy="25"
        fill="none"
        r="20"
        stroke="#888"
        stroke-dasharray="50 155"
        stroke-linecap="round"
        stroke-width="5"
      />
    </svg>
  </span>
</div>
`;
